### PR TITLE
[PDS-126257] Turning Tables: Checked Rejection Executor Services

### PR DESCRIPTION
**Pull request by jeremyk-91**
**Goals (and why)**:
- See PDS-126257
- We can't trust developers (including myself) to correctly remember to catch RejectedExecutionException: it's a random case that only happens occasionally at our biggest deployments.

**Implementation Description (bullets)**:
- A {@link CheckedRejectionExecutorService} is an executor where submitters of tasks MUST be prepared to handle execution being rejected - and we thus throw a checked exception when submitting tasks. (Thanks to @j-baker for the idea!)
- This should make things fail at compile time.

**Testing (What was existing testing like?  What have you done to improve it?)**:
- Added tests for the new classes. Otherwise the compiler should be as good a test as any that wiring is correct.

**Concerns (what feedback would you like?)**:
- Did I fudge the wiring up nonetheless?

**Where should we start reviewing?**: CheckedRejectionExecutorService.java

**Priority (whenever / two weeks / yesterday)**: next week?

### CODE REVIEW

File: timelock-agent/src/main/java/com/palantir/atlasdb/timelock/paxos/TimeLockPaxosExecutors.java [hunk 1]
```diff
@@ -39,15 +40,17 @@ private TimeLockPaxosExecutors() {
      *
      * It is assumed that tasks run on the local node will return quickly (hence the use of the direct executor).
      */
-    static <T> Map<T, ExecutorService> createBoundedExecutors(
-            MetricRegistry metricRegistry, LocalAndRemotes<T> localAndRemotes, String useCase) {
+    static <T> Map<T, CheckedRejectionExecutorService> createBoundedExecutors(
+            LocalAndRemotes<T> localAndRemotes, String useCase) {
         int numRemotes = localAndRemotes.remotes().size();
```
Comment (sudiksha27) [inline timelock-agent/src/main/java/com/palantir/atlasdb/timelock/paxos/TimeLockPaxosExecutors.java:45]: 👍

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/AutobatchingPaxosAcceptorNetworkClientFactory.java [hunk 2]
```diff
@@ -28,6 +28,7 @@
 import com.google.common.util.concurrent.ListenableFuture;
```
Comment (sudiksha27) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/AutobatchingPaxosAcceptorNetworkClientFactory.java:28]: nit: there are unused imports in some other places as well where ExecutorService is replaced by CheckedRejectionExecutorService

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosQuorumCheckingCoalescingFunction.java [hunk 3]
```diff
@@ -111,7 +112,7 @@ private PaxosQuorumCheckingCoalescingFunction(
     public static <REQ, RESP extends PaxosResponse, SERVICE, F extends CoalescingRequestFunction<REQ, RESP>>
     PaxosQuorumCheckingCoalescingFunction<REQ, RESP, F> wrapWithRemotes(
             List<SERVICE> services,
-            Map<SERVICE, ExecutorService> executors,
+            Map<SERVICE, CheckedRejectionExecutorService> executors,
             int quorumSize,
             Function<SERVICE, F> functionFactory) {
         List<FunctionAndExecutor<F>> functionsAndExecutors = KeyedStream.of(services)
```
Comment (sudiksha27) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosQuorumCheckingCoalescingFunction.java:118]: Where is this used?

Comment (jeremyk-91) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosQuorumCheckingCoalescingFunction.java:118]: It's used inside `wrap()`. I think you might be referencing the private constructor on line 68 though, which is unused? I've removed that.

Comment (jeremyk-91) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosQuorumCheckingCoalescingFunction.java:118]: Ah, I guess it's not. Well spotted. I'll remove this.

**Pull request merged**